File: sol/1415/1415.cpp
struct Solution {
    string getHappyString(int n, int k) {
        string res;
        int l = 1 << (--n);
        if (k > l * 3){
            return res;
        }
        int q = (--k) / l;
        k &= l - 1;
        res.push_back('a' + q);
        l >>= 1;
        while (n-- > 0){
            if (k & l){
                q = (q == 2) ? 1 : 2;
            } else {
                q = (q == 0) ? 1 : 0;
            }
            res.push_back('a' + q);
            l >>= 1;
        }
        return res;
    }
};

File: sol/1415/1415.py
class Solution:
    CHARS = 'abc'
    def getHappyString(self, n: int, k: int) -> str:
        l = 1 << (n - 1)
        if 3 * l < k:
            return ''
        res = ''
        q, k = divmod(k - 1, l)
        res += self.CHARS[q]
        l >>= 1
        for i in range(n - 1):
            if k & l:
                q = 1 if q == 2 else 2
            else:
                q = 1 if q == 0 else 0
            res += self.CHARS[q]
            l >>= 1
        return res
